### PR TITLE
GLOB-38999: Upgrade library to use apollo server v2

### PR DESCRIPTION
**Pull request by mlcohen**
Ticket: https://globality.atlassian.net/browse/GLOB-38999

Updated this package to use `apollo-server` v2. This update does not break the API or semantics. Everything will continue to work as before. Only internal changes were done.

Main changes:

1. updated `routes.graphql` to use new `ApolloServer` class
2. updated `routes.graphiql` to use the new GraphQL playground app

For (1), some changes had to be made to accommodate how a `ApolloServer` instance is created. For (2), graphiql has been replaced with a newer GQL tool called GraphQL playground (https://github.com/prisma-labs/graphql-playground)

Changes to this package has been tested against the styx GW. Everything continues to work as expected.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,42 +1,44 @@
...
+        "clean": "rm -rf ./lib",
         "build": "babel src --out-dir lib --ignore '**/__tests__/*,**/__mocks__/*'",
         "lint": "eslint src --cache",
+        "pack:dev": "./scripts/pack-dev.sh",
         "test": "jest",
         "verify": "yarn lint && yarn test"
     },
     "dependencies": {
-        "apollo-server-express": "^1.3.3",
+        "apollo-server-express": "^2.9.9",
```
Comment (mlcohen) [inline package.json]: Tada!

File: scripts/pack-dev.sh [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+yarn clean 
+yarn build
+
+rm globality-nodule-graphql-*.tgz
+yarn pack
+
+file=$(ls globality-nodule-graphql-*.tgz)
+name=${file%.*}
+
+mv $file $name.$(date +%s).tgz
```
Comment (mlcohen) [inline scripts/pack-dev.sh]: little convenience scripts to generate a packed library. Name of the package is suffixed with a unique timestamp.

File: src/__tests__/app.js [hunk 3]
```diff
@@ -19,14 +19,14 @@ export default function createApp() {
 
     express.get('/health', health);
 
-    express.use(
+    express.post(
```
Comment (mlcohen) [inline src/__tests__/app.js]: Needed to change `use` to `post` for apollo server v2 to work

File: package.json [hunk 4]
```diff
@@ -1,42 +1,44 @@
...
+        "graphql-playground-middleware-express": "^1.7.12",
         "helmet": "^3.12.0",
         "http-status-codes": "^1.3.0",
         "is-uuid": "^1.0.2",
         "jsonwebtoken": "^8.2.0",
         "lodash": "^4.17.5",
         "node-spookyhash-v2": "^1.0.1",
         "throat": "^4.1.0",
-        "uuid": "^3.2.1",
-        "yarn": "^1.5.1"
```
Comment (mlcohen) [inline package.json]: `yarn` was a dependency 😵

File: package.json [hunk 5]
```diff
@@ -54,7 +56,7 @@
         "eslint": "^4.18.2",
         "eslint-config-airbnb-base": "^12.1.0",
         "eslint-plugin-import": "^2.9.0",
-        "graphql": "^0.13.2",
+        "graphql": "^14.5.8",
```
Comment (mlcohen) [inline package.json]: This looks like a big jump going from `0.13` to `14.0` but it actually represents a single major version change

File: CHANGELOG.md [hunk 6]
```diff
@@ -0,0 +1,7 @@
+# Changelog
+
+### v0.32.0
+
+- Upgrade `apollo-server` depedency. Now using v2.
+- Upgrade `graphql` depencency. Now using v14.
+- `routes.graphiql` now uses [graphql playground](https://github.com/prisma-labs/graphql-playground)
```
Comment (mlcohen) [inline CHANGELOG.md]: Added a changelog file! 🤗

File: package.json (hunk 5 above)
Comment (cdm-ium) [reply package.json]: good ol' js.

File: src/routes/graphiql.js [hunk 7]
```diff
@@ -1,25 +1,30 @@
...
+                'request.credentials': 'include',
+            },
+        })(req, res, next);
+    };
 }
 
 
 setDefaults('routes.graphiql', {
-    endpointUrl: '/content/graphql',
+    endpointUrl: '/gql/graphql',
```
Comment (cdm-ium) [inline src/routes/graphiql.js]: I don't think you want to change this. At the very least, if we change this default, we'll need to change stuff in ansible too. https://globalityinc.slack.com/archives/CLR6QA1TJ/p1571854964190000

Comment (mlcohen) [reply src/routes/graphiql.js]: yep, good point. Will revert.

File: package.json [hunk 8]
```diff
@@ -1,42 +1,44 @@
...
     "name": "@globality/nodule-graphql",
-    "version": "0.31.0",
+    "version": "0.32.0",
     "description": "Node GraphQL Conventions",
     "main": "lib/",
     "repository": "git@github.com:globality-corp/nodule-graphql",
     "author": "Globality Engineering",
     "license": "MIT",
     "scripts": {
+        "clean": "rm -rf ./lib",
```
Comment (GabeLambda) [inline package.json]: `-f`? Bold.

File: src/routes/graphiql.js (hunk 7 above)
Comment (cdm-ium) [reply src/routes/graphiql.js]: Don't see an update here yet. I'm gonna sign off, but assuming someone can overwrite my review, feel free to have someone else approve!

File: package.json (hunk 8 above)
Comment (mlcohen) [reply package.json]: I mean, it's the same thing as `rimraf`

File: src/routes/graphiql.js (hunk 7 above)
Comment (mlcohen) [reply src/routes/graphiql.js]: yeah, I have a local change. Testing.

Comment (mlcohen) [reply src/routes/graphiql.js]: updated

File: src/__tests__/routes.test.js [hunk 9]
```diff
@@ -145,7 +145,7 @@ describe('routes', () => {
         });
         expect(response.body.errors).toHaveLength(1);
         expect(response.body.errors[0].extensions).toEqual(expect.objectContaining({
-            code: 503,
+            code: 'INTERNAL_SERVER_ERROR',
```
Comment (GabeLambda) [inline src/__tests__/routes.test.js]: Why the change here?

File: src/routes/graphql.js [hunk 10]
```diff
@@ -83,20 +81,13 @@ function formatError(error) {
...
-
-    return function configure(req) {
-        return {
-            cacheControl: config.routes.graphql.cacheControl,
-            context: req,
-            // merge in response extensions
-            formatResponse: injectExtensions(req),
-            rootValue: null,
-            schema,
-            tracing: config.routes.graphql.cacheControl || req.headers['x-request-trace'],
```
Comment (GabeLambda) [inline src/routes/graphql.js]: What happened to this?

Comment (mlcohen) [reply src/routes/graphql.js]: `tracing` was used for the apollo engine proxy that apollo-server v1 used to collect metrics. From the apollo-server v2 docs:

> If set to true, adds tracing or cacheControl metadata to the GraphQL response. This is primarily intended for use with the deprecated Engine proxy.  cacheControl can also be set to an object to specify arguments to the apollo-cache-control package, including defaultMaxAge, calculateHttpHeaders, and stripFormattedExtensions

I checked `styx`'s configuration to see if `routes.graphql.cacheControl` is set in AWS. it's not. If you feel like we should keep this I can update. This might be something we can add back later on when we have a better understand of how to use this specific feature. However, from what I'm reading, this is pretty much used by the Apollo Engine

File: src/__tests__/routes.test.js (hunk 9 above)
Comment (mlcohen) [reply src/__tests__/routes.test.js]: This is what apollo-server v2 now returns as internal server error code. Looking over this change, yeah, I can see this being a potential breaking change if any of our FE apps were expecting a `503`. I don't believe that's the case right now. For instance, phoenix does not currently check GraphQL errors for a `503`. With that being said, I can update the logic to continue returning `503` if an `INTERNAL_SERVER_ERROR` is detected.
